drpcmetadata/metadata.go

```diff
@@ -103,11 +112,57 @@ func Get(ctx context.Context) (map[string]string, bool) {
 }
 
 // GetValue retrieves a specific value by key from the context's metadata.
+// The input key is assumed to be lowercase. If metadata was created using
+// the provided helper functions (Add, AddPairs, NewIncomingContext, etc.),
+// this performs a fast O(1) lookup. If metadata was attached externally with
+// mixed-case keys, a slower fallback search is performed.
 func GetValue(ctx context.Context, key string) (string, bool) {
-	metadata, ok := Get(ctx)
+	metadata, ok := ctx.Value(metadataKey{}).(map[string]string)
 	if !ok {
 		return "", false
 	}
-	val, ok := metadata[key]
-	return val, ok
+	if val, ok := metadata[key]; ok {
+		return val, true
+	}
+	// TODO: Check if we really need this. Keeping this for now, to conform to
+	// grpc metadata semantics
+	for k, v := range metadata {
+		// We need to manually convert all keys to lower case,
+		// because metadata is a map,
+		//and there's no guarantee that the metadata
+		// attached to the context is created using our helper functions.
+		if len(k) == len(key) && strings.ToLower(k) == key {
+			return v, true
+		}
+	}
+	return "", false
+}
+
+// FastFromIncomingContext is a specialization of Get() and is
+// based on grpcutil.FastFromIncomingContext from the cockroach repo.
+// It extracts the metadata from the context, if any, by reference.
+//
+//   - Unlike Get, this variant does not guarantee that all the metadata keys
+//     are lowercase.
+//   - The caller promises to not modify the returned metadata -- the dRPC
+//     APIs assume that the map in the context remains constant.
+func FastFromIncomingContext(ctx context.Context) (map[string]string, bool) {
+	metadata, ok := ctx.Value(metadataKey{}).(map[string]string)
+	if !ok {
+		return nil, false
+	}
+	return metadata, true
+}
+
+// NewOutgoingContext attaches new metadata onto an outgoing context and returns
+// the context. Same as NewIncomingContext for now,
+// as we don't have separate keys for incoming and outgoing metadata.
+// Will be fixed as part https://github.com/cockroachdb/cockroach/issues/156444
+func NewOutgoingContext(ctx context.Context,
+	metadata map[string]string) context.Context {
+	newMetadata := make(map[string]string)
+	for k, v := range metadata {
+		newMetadata[strings.ToLower(k)] = v
+	}
+	return context.WithValue(ctx, metadataKey{}, newMetadata)
```

Comment:
I think this change should go in a separate PR. And how much effort would be to separate the keys and close the referenced issue since we are at it.

Reply:
Separating keys is not in the scope of these changes and it's not clear at this point if I'd take them up, but I'd prefer to add this method here with a note that it is WIP, since it helps keep the semantics in the dependent calling code clear. I have another draft PR with the changes in the calling code which I will share soon.